### ISSUE TITLE
Bug: Update of assigns used in for loop to create rows in a <table> does not work properly

### ISSUE THREAD

**Issue by steven-cole-elliott** [grych/drab#50]
Say I have just created a new Phoenix 1.3 project with Drab v0.5.4 and have generated an `Accounts` context with the following `user.ex` struct:
```
schema "users" do
    field :email, :string
    field :name, :string

    timestamps()
end
```

In a corresponding `user_controller.ex`, I have a standard `index` action that renders the `index.html.drab` template, passing it a `users` assigns which is a list of users:

```
def index(conn, _params) do
    users = Accounts.list_users()
    changeset = Accounts.change_user(%User{})

    render(conn, "index.html", users: users, changeset: changeset)
end
```

The `index.html.drab` displays the list of users in the `users` assigns inside a `<tbody>` element, rendering a new `<tr>` for each user in the `users` assigns, and new `<td>` for both the `name` and `email` fields on that user:

```
<table class="table">
  <thead>
    <tr>
      <th>Name</th>
      <th>Email</th>

      <th></th>
    </tr>
  </thead>
  <tbody>
<%= for user <- @users do %>
    <tr>
      <td><%= user.name %></td>
      <td><%= user.email %></td>

      <td class="text-right">
        <span><%= link "Show", to: user_path(@conn, :show, user), class: "btn btn-default btn-xs" %></span>
        <span><%= link "Edit", to: user_path(@conn, :edit, user), class: "btn btn-default btn-xs" %></span>
        <span><%= link "Delete", to: user_path(@conn, :delete, user), method: :delete, data: [confirm: "Are you sure?"], class: "btn btn-danger btn-xs" %></span>
      </td>
    </tr>
<% end %>
  </tbody>
</table>

<%= render "form.html", Map.put(assigns, :action, user_path(@conn, :create)) %>
```

To demonstrate the issue, notice the call to render a form to create a new user on the same `index.html.drab` template above on the last line of the snippet.  

Here is the `form.html.eex` template that is rendered within the `index.html.drab` template:

```
<%= form_for @changeset, @action, fn f -> %>
  <%= if @changeset.action do %>
    <div class="alert alert-danger">
      <p>Oops, something went wrong! Please check the errors below.</p>
    </div>
  <% end %>

  <div class="form-group">
    <%= label f, :name, class: "control-label" %>
    <%= text_input f, :name, class: "form-control" %>
    <%= error_tag f, :name %>
  </div>

  <div class="form-group">
    <%= label f, :email, class: "control-label" %>
    <%= text_input f, :email, class: "form-control" %>
    <%= error_tag f, :email %>
  </div>

  <div class="form-group" drab-click="create">
    <button type="button">
      Submit
    </button>
  </div>
<% end %>
```

I've placed a `drab-click` on the button in the form to hit the `create` function in the `user_commander.ex` Drab commander, which will create the user using the name and email from the form, and then `poke` the `users` assigns that is used to render the table of users with the new list of users:

```
def create(socket, sender) do

    user_params = sender.params["user"]

    case DrabTest.Accounts.create_user(user_params) do
      {:ok, user} ->
        users = DrabTest.Accounts.list_users()
        poke socket, users: users
      {:error, %Ecto.Changeset{} = changeset} ->
        poke socket, test_text: "Error. Could not create user."
    end

end
```

However, when I click the Submit button, the new user is created successfully, but the existing `<tbody>` is not updated, and instead, the new list of users poked in the `users` assigns is rendered above the containing `<table>` element in a `<span drab-ampere='...'>` element, without the `<tr>` and `<td>` elements.  Here is a snippet of the resulting HTML:

```
<span drab-partial="haztsnbvhe3ta">
<h2>Listing Users</h2>

<span drab-ampere="gqzdsnjqge3tm">
    
    
      John Smith
      john.smith@test.com

      
        <span><span drab-ampere="gi4domjtgiyds"><a class="btn btn-default btn-xs" href="/users/2">Show</a></span></span>
        <span><span drab-ampere="giydcmzrhazde"><a class="btn btn-default btn-xs" href="/users/2/edit">Edit</a></span></span>
        <span><span drab-ampere="geytgojtgq2deni"><a class="btn btn-danger btn-xs" data-confirm="Are you sure?" data-csrf="PCYcYAUaICkEJwsUO1Q1ARAvLwM6AAAAEAz1nxGDRbMLBgQFuly5wQ==" data-method="delete" data-to="/users/2" href="#" rel="nofollow">Delete</a></span></span>
      
    

    
      Test User
      testuser@test.com

      
        <span><span drab-ampere="gi4domjtgiyds"><a class="btn btn-default btn-xs" href="/users/6">Show</a></span></span>
        <span><span drab-ampere="giydcmzrhazde"><a class="btn btn-default btn-xs" href="/users/6/edit">Edit</a></span></span>
        <span><span drab-ampere="geytgojtgq2deni"><a class="btn btn-danger btn-xs" data-confirm="Are you sure?" data-csrf="PCYcYAUaICkEJwsUO1Q1ARAvLwM6AAAAEAz1nxGDRbMLBgQFuly5wQ==" data-method="delete" data-to="/users/6" href="#" rel="nofollow">Delete</a></span></span>
      
    
</span><table class="table">
  <thead>
    <tr>
      <th>Name</th>
      <th>Email</th>

      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>John Smith</td>
      <td>john.smith@test.com</td>

      <td class="text-right">
        <span><span drab-ampere="gi4domjtgiyds"><a class="btn btn-default btn-xs" href="/users/2">Show</a></span></span>
        <span><span drab-ampere="giydcmzrhazde"><a class="btn btn-default btn-xs" href="/users/2/edit">Edit</a></span></span>
        <span><span drab-ampere="geytgojtgq2deni"><a class="btn btn-danger btn-xs" data-confirm="Are you sure?" data-csrf="ZVcBEnorQn1iTCtVBwABGwtvGSAzAAAAR8T+4S1IVgCzLx60j7TwbA==" data-method="delete" data-to="/users/2" href="#" rel="nofollow">Delete</a></span></span>
      </td>
    </tr>

  </tbody>
</table>
``` 

What I would expect is for the poke of `:users` to update the `<tbody>`, not inject the updated `:users` assigns in a `<span>` above the `<table>` element.   I don't believe it is an issue with using Drab with `for` inside a template, as the demo page shows a simple implementation of updating a list of users.  

I've pasted some screenshots before and after creating a new user to provide a visual as well.

### Before Creating Test User
![before creating test user](https://user-images.githubusercontent.com/24230004/29690997-feb07c14-88ee-11e7-84a3-c3e86d7a505f.png)

### After Creating Test User
![after creating test user](https://user-images.githubusercontent.com/24230004/29690996-feae8846-88ee-11e7-9962-eb5381f2f10e.png)

**Comment by grych** [comment]
This is a fundamental issue with the html tables. Thanks for detailed report!

Drab surrounds the expression with `<span drab-ampere=''>` to find out where to poke the updated assigns. 

But browsers don't allow `<span>` to be inside the table. Consider the simple example in plain html:

```html
<table>
  <span>anything</span>
</table>
```

When you inspect this in the browser, looks like it is pushed away from the table:

```html
<span>anything</span>
<table>
  
</table>
```

The solution (in your example) would be to drab `<tbody>`, rather than injecting `<span>`. I am now rewriting the eex engine and will try to fix this one.

But since it is fixed, unfortunatelly **Drab.Live does not work with html tables**

**Comment by justincjohnson** [comment]
Forgive my ignorance but is there a reason drab can't put its drab-ampere attributes right on the dom element instead of wrapping in new span elements?

**Comment by grych** [comment]
Actually, there are reasons: 1) this is much simpler solution and 2) there is an issue when rendering the partial with text only, without any tags.

1) Consider:

```html1
<tag1>
  Some text <%= @assign %> and <%= @assign2 %>
</tag1>
```

In case when `tag1` is drabbed with `drab-ampere` attribute, Drab needs to remember the whole text pattern with two expressions inside. 

Injecting the span it the simplest solution. But, while implementing it, I did not know about the table issue. So it must be changed to work in that way.

Acutally, there is such solution implemented in drab - in attributes, scripts and textareas. There is no way to put the `<span>` inside the attribute value :) But it is under reconstruction because it does not work properly (#49).

2) Consider the partial:

```html
User name: <%= @user.name%> surname: <%= @user.last_name %>
```

which is rendered in some template:

```html
<%= for user <- @users do %>
  <td><%= render "user.html", user: user %></td>
<% end %>
```

At the compile time, Drab does not know that `user.html` will be rendered inside the `<td>` tag. And in this partial, there is no tag to connect to!

The solution would be to force developer to always put any tag in the partial. In the other words, `user.html` from the above will not compile, you will be forced to change your code to:

```html
<td>User name: <%= @user.name%> surname: <%= @user.last_name %></td>
```

```html
<%= for user <- @users do %>
  <%= render "user.html", user: user %>
<% end %>
```

**Comment by justincjohnson** [comment]
That makes sense.  Thanks for the clarification.  It does sound to me like this needs a more complex virtual dom that can keep track of where content needs to get injected without having to impose itself on the structure of the rendered dom.  I'm loving the overall concept for drab and _really_ wanting it to work smoothly for all these edge cases.  :)

Another similar problem we ran into is that the injected spans can mess up styles when say you're using some existing css framework like material design that requires that two different dom elements come right after one another.

**Comment by grych** [comment]
@steven-cole-elliott 
Could you please re-test with current master?

**Comment by gen1321** [comment]
@grych works good for me

**Comment by grych** [comment]
Fixed in v0.6.0-pre.1
